agentops_property - internal objects tracking the right way

teocns opened the pull request:
Introduces the  `agentops_property` descriptor for safe and consistent objects tracking;


- The `check_call_stack_for_agent_id` function has been refactored to utilize this property descriptor improving the process of identifying agent IDs from the call stack. 

- Property access nomenclature is standardized to `_agentops_<name>` for _private attrs_ and `agentops_<name>` for _public attrs_

teocns committed:
AgentProperty descriptor

The function `check_call_stack_for_agent_id` is designed to search
through the call stack to find an instance of a class that contains
specific `AgentProperty` descriptors, particularly looking for an
`agent_ops_agent_id`. Here's a step-by-step explanation:1. **Function
Definition and Docstring**:   - The function is defined to return either
a `UUID` or `None`.   - The docstring explains that it looks through the
call stack for the class that called the LLM (Language Learning Model)
and checks for `AgentProperty` descriptors.2. **Inspecting the Call
Stack**:   - The function uses `inspect.stack()` to get the current call
stack, which is a list of `FrameInfo` objects representing the stack
frames.3. **Iterating Through Stack Frames**:   - It iterates over each
`frame_info` in the call stack.4. **Accessing Local Variables**:   - For
each frame, it accesses the local variables using
`frame_info.frame.f_locals`.5. **Checking Each Local Variable**:   - It
iterates over each local variable (`var_name`, `var`) in the current
frame.6. **Stopping at Main**:   - If the variable name is `"__main__"`,
it returns `None` and stops further processing.7. **Checking for
AgentProperty Descriptors**:   - It tries to check if the variable
(`var`) has `AgentProperty` descriptors.   - It gets the type of the
variable (`var_type`).   - It retrieves all class attributes of
`var_type` into a dictionary `class_attrs`.8. **Looking for Specific
Descriptors**:   - It looks for an attribute named `agent_ops_agent_id`
in `class_attrs`.   - If `agent_ops_agent_id` is an instance of
`AgentProperty`, it retrieves the agent ID using
`agent_id_desc.__get__(var, var_type)`.9. **Returning the Agent ID**:
- If an agent ID is found, it optionally retrieves the agent name using
a similar process and returns the agent ID.   - If no agent ID is found,
it continues to the next variable.10. **Handling Exceptions**:    - If
any exception occurs during the process, it catches the exception and
continues to the next variable.11. **Returning None**:    - If no agent
ID is found in the entire call stack, it returns `None`.This function is
useful for debugging or tracking purposes, where you need to identify
the agent that initiated a particular call in a complex system.

Signed-off-by: Teo <teocns@gmail.com>

areibman commented on agentops/decorators.py:
```diff
@@ -311,30 +312,36 @@ def sync_wrapper(*args, session: Optional[Session] = None, **kwargs):
+                        agent_id=self._agentops_agent_id,
                         session=session,
                     )
                 except AttributeError as e:
```
I'm not sure what this bot is talking about lol

areibman commented on agentops/descriptor.py:
I'm not entirely sure I understand the purpose of this file. Why are we replacing the helper function with `AgentOpsDescriptor`?

areibman commented on agentops/helpers.py:
```diff
@@ -99,22 +106,7 @@ def remove_unwanted_items(value):
-                )
-                return getattr(var, "agent_ops_agent_id")
-    return None
+    return AgentOpsDescriptor.from_stack()
```
Everywhere in this PR we instantiate `AgentOpsDescriptor` inline instead of having class instance variables. Why are we doing this instead of just running `from_stack` as a separately defined helper function?

teocns replied:
The signature would've better been off `AgentOpsDescriptor.get_from_stack(property="agent_id")`

Reason you want a descriptor is because you need a distinguishable way of identifying your own properties (agentops props)

> Everywhere in this PR we instantiate AgentOpsDescriptor inline instead

...it's not being instantiated, it's a static method being invoked, as if it was a function under a module named `AgentOpsDescriptor`. 

Most of the "why" are answered with behavior encapsulation

teocns commented:
The previous helper was just a helper in the middle of a `helpers` utility file. This introduces a new concept that's quite fundamental for what's going on in the AgentOps library: mocking / patching, property lookup, etc. Challenge being faced right now is tracking of _AgentOps'_ own properties installed onto objects. 

For instance, say you wanna lookup `_agentops_agent_id` onto a property.

```

class Lies:

    def __init__(self):
        self._agentops_agent_id = uuid4()
      

    def __getattr__(self):
      return "lied"



Lies()._agentops_agent_id # lied
```

Without a descriptor, there is no way to determine whether the value accessed at `_agentops_agent_id` is genuine, which is exactly what was happening with retrieving the agent_id from call stack: a sea of lies.

areibman commented:
> The previous helper was just a helper in the middle of a `helpers` utility file. This introduces a new concept that's quite fundamental for what's going on in the AgentOps library: mocking / patching, property lookup, etc. Challenge being faced right now is tracking of _AgentOps'_ own properties installed onto objects.
> 
> For instance, say you wanna lookup `_agentops_agent_id` onto a property.
> 
> ```
> 
> class Lies:
> 
>     def __init__(self):
>         self._agentops_agent_id = uuid4()
>       
> 
>     def __getattr__(self):
>       return "lied"
> 
> 
> 
> Lies()._agentops_agent_id # lied
> ```
> 
> Without a descriptor, there is no way to determine whether the value accessed at `_agentops_agent_id` is genuine, which is exactly what was happening with retrieving the agent_id from call stack: a sea of lies.

Great catch.

teocns committed:
fmt

Signed-off-by: Teo <teocns@gmail.com>

teocns commented:
I can't get `black` to pass, do we have a pre-commit hook to run? might aswell be `black` version/rules mismatches: how are we standardizing that across development environments?

dot-agi commented:
@teocns @areibman is this good to merge? Will be helpful for #500.

areibman commented:
Not good to merge. This is failing our integration with CrewAI:
```
(env) (base) ➜  job-posting git:(main) ✗ python main.py
🖇 AgentOps: Session Replay: https://app.agentops.ai/drilldown?session_id=fd72ba4e-8d25-43e4-97b9-dd0fb88baf91
What is the company description?
job
What is the company domain?
jobby
What are the hiring needs?
jobbb
What are specific_benefits you offer?
good job
🖇 AgentOps: Failed to track an agent with the @track_agent decorator.
🖇 AgentOps: Failed to track an agent with the @track_agent decorator.
🖇 AgentOps: Failed to track an agent with the @track_agent decorator.
```

Here's a good example repo to use. Use the job posting example, I find it easiest
https://github.com/AgentOps-AI/crewAI-examples

dot-agi commented:
@teocns I think you should update the examples with the `track_agent` decorator to demonstrate its usage and that would also help test it out for those agents frameworks.

braelyn-ai commented:
> I can't get `black` to pass, do we have a pre-commit hook to run? might aswell be `black` version/rules mismatches: how are we standardizing that across development environments?

black is super opinionated and has no configuration. you can run it locally to see what its upset about

teocns committed:
agentops_property, __set_name__ to automatically handle attribute naming

1. Renamed AgentOpsDescriptor to agentops_property;
2. Thanks to __set_name__, The descriptor will now automatically know
its own name when it's assigned as a class attribute

Signed-off-by: Teo <teocns@gmail.com>

teocns committed:
test: update agent property tests, add more coverage on call stacks

teocns commented:
## Key Changes
1. Improved property name resolution:
   - Added fallback mechanisms when `__set_name__` isn't called
   - Better handling of dynamically added properties
   - Proper handling of `agentops_` prefix stripping

2. Enhanced error handling:
   - Graceful handling of property name resolution failures
   - Better exception handling in stack inspection
   - Clear error messages when property names can't be determined

3. Added comprehensive test suite:
   - Tests for basic property get/set operations
   - Tests for stack frame inspection
   - Tests for inheritance scenarios
   - Tests for error cases and edge conditions

## Technical Details
- The descriptor now properly handles both static class attribute definition and dynamic property addition
- Stack frame inspection is more resilient to exceptions
- Property naming is more consistent with automatic prefix handling
- Added proper type hints and documentation

## Testing
Added comprehensive test suite covering:
- Direct property access
- Nested function calls
- Multiple agents in different stack frames
- Inheritance scenarios
- Error conditions
- Dynamic property addition

teocns committed:
black

Signed-off-by: Teo <teocns@gmail.com>

teocns committed:
refactor(decorators): using public name accessor

teocns committed:
Big W

teocns committed:
Merge branch 'main' into agent-property-descriptor

teocns commented:
Big Win. It now comes with general pydantic support 🥳 

For CrewAI, it requires the [following patch](https://gist.github.com/teocns/517435816becc1581250d145bd9eead9); here's why:

For clean code reasons I renamed the property prefix standard to `agentops` (unified), rather than `agent_ops`
examples:
```
agentops_agent_id
_agentops_agent_name
```
Which makes a lot more sense to contain the scope and purpose within a single snake case group `_scope_`

@areibman let's merge this  - let me know if you wanna move back to agent_ops property nomenclature for an express merge

teocns committed:
ruff

Signed-off-by: Teo <teocns@gmail.com>

areibman commented:
Amazing, works for me on several CrewAI and Autogen examples

areibman reviewed (approved):
Great stuff

teocns commented:
@the-praxs  second pair of eyes before merge pls?

dot-agi reviewed (approved):
LGTM 🚀
